File: src/main/kotlin/top_150/_36_valid_sudoku.java
package top_150;

public class _36_valid_sudoku {
    public boolean isValidSudoku(char[][] board) {
        // try to come up with an idea that I check all the groups or sections
        // by 9 times
        // sum from 1 through 9 is 45
        for (int i = 0; i < board.length; i++) {
            boolean[] found = new boolean[27];

            for (int j = 0; j < 9; j++) {
                // 1. horizontal (column by column) check
                char hNum = board[j][i];
                if ('1' <= hNum && hNum <= '9') {
                    int idx = hNum - '1';
                    if (found[idx]) {
                        return false;
                    } else {
                        found[idx] = true;
                    }
                }
                // 2. vertical (row by row) check
                char vNum = board[i][j];
                if ('1' <= vNum && vNum <= '9') {
                    int idx = vNum - '1' + 9;
                    if (found[idx]) {
                        return false;
                    } else {
                        found[idx] = true;
                    }
                }

                var r = j / 3 + (i / 3) * 3;
                var c = j % 3 + (i % 3) * 3;

                var num = board[r][c];
                if ('1' <= num && num <= '9') {
                    int idx = num - '1' + 18;
                    if (found[idx]) {
                        return false;
                    } else {
                        found[idx] = true;
                    }
                }
            }
        }
        return true;
    }

    public static void main(String[] args) {
        var s = new _36_valid_sudoku();
        System.out.println(s.isValidSudoku(new char[][]{
                {'5','3','.','.','7','.','.','.','.'},
                {'6','.','.','1','9','5','.','.','.'},
                {'.','9','8','.','.','.','.','6','.'},
                {'8','.','.','.','6','.','.','.','3'},
                {'4','.','.','8','.','3','.','.','1'},
                {'7','.','.','.','2','.','.','.','6'},
                {'.','6','.','.','.','.','2','8','.'},
                {'.','.','.','4','1','9','.','.','5'},
                {'.','.','.','.','8','.','.','7','9'},
        }));
        System.out.println(s.isValidSudoku(new char[][]{
                {'8','3','.','.','7','.','.','.','.'},
                {'6','.','.','1','9','5','.','.','.'},
                {'.','9','8','.','.','.','.','6','.'},
                {'8','.','.','.','6','.','.','.','3'},
                {'4','.','.','8','.','3','.','.','1'},
                {'7','.','.','.','2','.','.','.','6'},
                {'.','6','.','.','.','.','2','8','.'},
                {'.','.','.','4','1','9','.','.','5'},
                {'.','.','.','.','8','.','.','7','9'},
        }));
        System.out.println(s.isValidSudoku(new char[][]{
                {'7','.','.','.','4','.','.','.','.'},
                {'.','.','.','8','6','5','.','.','.'},
                {'.','1','.','2','.','.','.','.','.'},
                {'.','.','.','.','.','9','.','.','.'},
                {'.','.','.','.','5','.','5','.','.'},
                {'.','.','.','.','.','.','.','.','.'},
                {'.','.','.','.','.','.','2','.','.'},
                {'.','.','.','.','.','.','.','.','.'},
                {'.','.','.','.','.','.','.','.','.'},
        }));
    }
}

File: src/main/kotlin/medium/_36_valid_sudoku.kt
package medium

class _36_valid_sudoku {
    fun isValidSudoku(board: Array<CharArray>): Boolean {
        val visited = mutableSetOf<Char>()
        for (i in 0 .. 8) {
            for (j in 0 .. 8) {
                val row = (j / 3) + (i / 3) * 3
                val column = (j % 3) + (i * 3) - ((i / 3) * 9)
                if (board[row][column] == '.') continue
                if (visited.contains(board[row][column])) {
                    return false
                } else {
                    visited.add(board[row][column])
                }
            }
            visited.clear()
        }

        for (i in 0 .. 8) {
            // by row
            for (j in 0..8) {
                println("row: $i, column: $j")
                if (board[i][j] == '.') continue
                if (visited.contains(board[i][j])) {
                    return false
                } else {
                    visited.add(board[i][j])
                }
            }
            visited.clear()

            // by column
            for (j in 0..8) {
                println("row: $j, column: $i")
                if (board[j][i] == '.') continue
                if (visited.contains(board[j][i])) {
                    return false
                } else {
                    visited.add(board[j][i])
                }
            }
            visited.clear()
        }
        return true
    }
}
